src/main/java/com/example/marspictmp/service/PictureService.java

```diff
@@ -0,0 +1,36 @@
+package com.example.marspictmp.service;
+
+import java.util.Comparator;
+import java.util.NoSuchElementException;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import com.example.marspictmp.controller.NasaFeignClient;
+
+import lombok.RequiredArgsConstructor;
+
+@Service
+@RequiredArgsConstructor
+public class PictureService {
+
+    @Value("${nasa.api_key}")
+    private String apiKey;
+
+    private final RestTemplate restTemplate;
+
+    private final NasaFeignClient nasaController;
+
+    @Cacheable("largest")
+    public byte[] getLargestPicture(Integer sol, String camera) {
+        var photosNodeResponse = nasaController.getNasaPhotos(apiKey, sol, camera);
+        return photosNodeResponse.findValues("img_src").parallelStream()
+          .map(url -> {
+              var picUrl = restTemplate.headForHeaders(url.asText()).getLocation();
+              return restTemplate.getForObject(picUrl, byte[].class);
```

Comment:
In that way, service will load every photo for specified sol. I think it's better to add an additional call for Headers using redirectUrl to get the actual content length. Then after comparing sizes we can call getForObject only once and decrease the network load. Other option is to configure RestTemplate using LaxRedirectStrategy to automatically redirect HEAD requests, in that case headForHeaders will return the actual content-length size, as it will execute an internal redirect.